feat: add support for beforeSend function to edit or drop events

wowrumal opened the pull request:
## :bulb: Motivation and Context
Required to filter events that the user doesn't want to send to the server.

## :green_heart: How did you test it?
Wrote a test + manual

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

marandaneto commented:
agree with `before_send` the way to go https://posthog.com/docs/libraries/js/features#amending-or-sampling-events

haritowa committed:
Introduce beforeSend hook

- Replace EventsSanitizer with beforeSend closure. Integrate the hook in all places before queue.add();
- Add test for the new beforeSend hook;
- Fix capture snapshots event check;

---------

Co-authored-by: Kirill Tsyvaka <wowrumal93@gmail.com>
Co-authored-by: haritowa <haritowa@gmail.com>

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -461,13 +461,12 @@ let maxRetryDelay = 30.0
-                properties: sanitizedProperties
-            ))
+            guard let event = buildEvent(event: "$identify", distinctId: distinctId, properties: properties) else {
+                return
```
we should add logging when events are dropped

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -771,13 +774,25 @@ let maxRetryDelay = 30.0
+            timestamp: timestamp
+        )
+
+        return config.beforeSend(event)
```
i think we should not run beforeSend for `$snapshot` events (just a hunch), we have to check the JS SDK and do the same if thats the case

marandaneto replied:
Better to check with replay team.
Dropped snapshots can cause a broken recording

haritowa replied:
We are ready to add this into PR, waiting for your clarification

haritowa replied:
I thought a bit more about this warning, and I think it still does not cover all the funky edge cases. For example, if the user modifies a snapshot event, we will not print any warnings because it is not `nil` after the mutation. This is consistent with the JS implementation, yet I think it is not the correct behavior.

The simplest way to fix it is to compare events before and after the hook and print a warning if:
-  the event changed and the original event was `knownUnsafeEditableEvent`
-  the event changed and the mutated event became `knownUnsafeEditableEvent`

Alternatively, we could introduce a separate hook for unsafe events, but again, this is a bit messy and also not aligned with the JS implementation.

Lastly, we can keep things as they are (print a warning only if the hook returns `nil` for unsafe events)

Also, we don't copy `PostHogEvent` either before or after the `hook` invocation. I am not sure if this is okay because before our change, the event object wasn't exposed to consumers (i.e., it wasn't possible to modify this object or capture it in the hook closure and modify it later).

We tested the "modified outside of hook closure" case, though, and PostHog actually sent the correct unmodified event; not sure why :D

haritowa replied:
Let's figure out the remaining questions in this thread:
> I thought a bit more about this warning, and I think it still does not cover all the funky edge cases.

https://github.com/PostHog/posthog-ios/pull/357#discussion_r2150169586

> Also, can we somehow allow editing the knownUnsafeEditableEvent list from the outside? This way, we can silence these warnings in case we know what we are doing. Our log console is a mess after this change :D

https://github.com/PostHog/posthog-ios/pull/357#discussion_r2150095809

Sorry for me being annoying, we want to finalize this contribution and use it in our release :)

haritowa replied:
Yeah, I think we will just live with these logs, I don't want to miss something important by silencing all the logs. And I agree, this feels wrong to expose internal stuff just to solve our minor issue

haritowa commented on PostHog/PostHogConfig.swift:
```diff
@@ -66,9 +68,13 @@ import Foundation
 
     /// Hook that allows to sanitize the event properties
     /// The hook is called before the event is cached or sent over the wire
+    @available(*, deprecated, message: "Use beforeSend instead")
```
We thought about some "seamless migration," i.e., to mark this prop as write-only and deprecated, and actually just set up a beforeSend closure with sanitizer logic. Yet, I think it is vague for the end user because they can override their property sanitizer when they start to use the beforeSend hook. I think the js lib has this prop deprecated(yet still available) too

haritowa commented on PostHog/PostHogConfig.swift:
```diff
@@ -6,6 +6,8 @@
 //
 import Foundation
 
+public typealias BeforeSendBlock = (PostHogEvent) -> PostHogEvent?
```
What should we do if the `beforeSend` hook received nil and returned non-nil? :D

Should we send it? Should we wrap `beforeSend` with a nil check and filter our such invocations?
If we introduce this signature, we have to handle such cases...

Actually, swift compiler will gladly allow us to pass `(T?) -> T?` to `(T) -> T?` prop, yet I am not sure how this will work for objc consumers

```swift
import Foundation

class DataClass {
  var prop: String

  init(prop: String) {
    self.prop = prop
  }
}

class Manager {
  var hook: (DataClass) -> DataClass? = { $0 }
}

func mutateFunction(input: DataClass?) -> DataClass? {
  input?.prop = "test 2"
  return input
}

let manager = Manager()
let initialDataClass = DataClass(prop: "test 1")

print(manager.hook(initialDataClass)?.prop) // Optional("test 1")
manager.hook = mutateFunction

print(manager.hook(initialDataClass)?.prop) // Optional("test 2")
```

haritowa replied:
Done, yet I am not sure how to test it with objc. Do you have some objc related tests?

haritowa replied:
Ah, I almost forgot. We actually got some cool Swift sugar, and a single hook closure can be set like [this](https://github.com/PostHog/posthog-ios/pull/357/files#diff-ccdae913e9a4a620023a87b8f54f7fb465761883177a4a7d91aeec9d584feb72R848) (even with our variadic implementation). I think this API is 🔥

```swift
config.setBeforeSend { $0.event == "first_event" ? nil : $0 }
```

Yet all of the below is valid too:
```swift
setBeforeSend { $0.event == "test_event" ? nil : $0 }

setBeforeSend(
    { $0.event == "other_event" ? nil : $0 },
    { $0.event = "modified_event"; return $0 }
)

setBeforeSend([
    { $0.event == "other_event" ? nil : $0 },
    { $0.event = "modified_event"; return $0 }
])
```

wowrumal committed:
Fixes according to git discussions

Add logging for dropped events. 
Add @objc property decorations. 
Add known events warning

haritowa commented:
☝️ we shipped some updates, could you please resolve already fixed comments so we can focus on the remaining bits

wowrumal committed:
Rename to PostHogKnownUnsafeEditableEvent. Add optional to param in BeforeSendBlock

haritowa committed:
Implement new hook setup interface: with support for multiple blocks and better objc interop

haritowa commented:
I think we should update PR title, it mutated into something completely different 😂

ioannisj reviewed (commented):
Thanks all — I think this is already in great shape. Really appreciate the effort and thought put into this! ❤️

Only thing left is to add a changelog entry so CI passes, and then I think we’re good to merge. 🚀

wowrumal committed:
Update PostHogTests/PostHogSDKTest.swift comment

Co-authored-by: Ioannis J <yiannis@posthog.com>

wowrumal committed:
Edit changelog
